src/SFML/Graphics/SpriteBatch.cpp

```diff
@@ -0,0 +1,267 @@
+                        const Transform& transform,
+                        float            depth)
+{
+    const bool unsupportedType = type != PrimitiveType::TriangleFan && type != PrimitiveType::TriangleStrip &&
+                                 type != PrimitiveType::Triangles;
+
+    if (unsupportedType)
+    {
+        err() << "SpriteBatch supports only triangle-based primitive types.";
+        return;
+    }
+
+    if (count == 0)
+        return;
+
+    // Split into triangles
+    switch (type)
+    {
+        case PrimitiveType::TriangleStrip:
+            for (std::size_t i = 2; i < count; i++)
+            {
+                pushTriangle(vertices[i - 2], vertices[i - 1], vertices[i], transform, texture, depth);
+            }
+            break;
+
+        case PrimitiveType::TriangleFan:
+            for (std::size_t i = 2; i < count; i++)
+            {
+                pushTriangle(vertices[0], vertices[i - 1], vertices[i], transform, texture, depth);
+            }
+            break;
+
+        case PrimitiveType::Triangles:
+            for (std::size_t i = 2; i < count; i += 3)
+            {
+                pushTriangle(vertices[i - 2], vertices[i - 1], vertices[i], transform, texture, depth);
+            }
+            break;
+
+        default:
+            // Shouldn't ever be hit
+            break;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::draw(RenderTarget& target, const RenderStates& states) const
+{
+    if (m_updateRequired)
+    {
+        updateBatch();
+        m_updateRequired = false;
+    }
+
+    RenderStates statesCopy(states);
+
+    std::size_t startTriangle = 0;
+
+    for (std::size_t i = 0; i < m_batches.size(); i++)
+    {
+        statesCopy.texture = m_batches[i].texture;
+
+        const auto& vertices = m_batchMode == BatchMode::Deferred ? m_unsortedVertices : m_vertices;
+
+        target.draw(&vertices[startTriangle], m_batches[i].vertexCount, PrimitiveType::Triangles, statesCopy);
+
+        startTriangle += m_batches[i].vertexCount;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::pushTriangle(const Vertex&    a,
+                               const Vertex&    b,
+                               const Vertex&    c,
+                               const Transform& transform,
+                               const Texture*   texture,
+                               float            depth)
+{
+    m_triangles.emplace_back(texture, depth);
+
+    m_unsortedVertices.emplace_back(transform * a.position, a.color, a.texCoords);
+    m_unsortedVertices.emplace_back(transform * b.position, b.color, b.texCoords);
+    m_unsortedVertices.emplace_back(transform * c.position, c.color, c.texCoords);
+
+    m_updateRequired = true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::updateBatch() const
+{
+    // Sanity check
+    if (m_triangles.size() == 0)
```

Comment:
```suggestion
    if (m_triangles.empty())
```

Reply:
Better yet, just use `assert` if this is something that you think will always be false. Debug builds will catch the error but release builds won't incur the performance penalty of this check.

Reply:
This one is a runtime check, although I'm not sure if it's needed given the method's logic.

Reply:
If you're not sure it's needed, then it's a great candidate for an assert. `assert`s are runtime checks.

Reply:
It's not needed in the sense that it may be redundant. `m_triangles` is supposed to have a varying size (including 0 triangles) 😅

Reply:
If it's perfectly valid to have zero size, then it's not a sanity check so you can remove that comment. To me, "sanity check" connotes code that is likely unnecessary. If you always need to handle that case, it's not a sanity check. It's just code that is required to make the implementation correct.